### ISSUE TITLE
XCode simulators failing when booting or using openurl

### ISSUE THREAD

**Issue by fpkamp** [actions/runner-images#7971]
### Description

Hi, I have been redirected from GitHub Support to describe our use case and perhaps influence future performance of your runner for macOS. 
We have a use case where we need to use XCode 14.3.1 and XCode 15.0 on the macOS13 runner. 
We create simulators with iPhones and navigate to a url and then we dispose the simulator.
All of the above works pretty well locally, but it fails when executing in GitHub Actions.
The behavior between XCode versions is different, with XCode 14 timing out on booting or navigating, however XCode 15 fails on other functions (perhaps an effect of changes to XCode itself) like binding launchd_sim.
An example failure message I get would be:
```
An error was encountered processing the command (domain=NSPOSIXErrorDomain, code=60):
Unable to boot the Simulator.
launchd failed to respond.
Underlying error (domain=com.apple.SimLaunchHostService.RequestError, code=4):
	Failed to start launchd_sim: could not bind to session, launchd_sim may have crashed or quit responding
```
In the ticket that has originally redirected me here, Arthur says he was successful with adding a cache clean (`~/Library/Caches/` and `~/Library/Developer/CoreSimulator/Caches/`) and waiting 60-120 seconds, however upon trying that I observed that it only works for a single simulator and subsequent simulators would fail (at least using a rinse and repeat approach).

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [X] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [X] macOS 13
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Image version and build link

version: 20230611.2
workflow run: https://github.com/fingerprintjs/fingerprintjs-pro/actions/runs/5646633166/job/15295123030

### Is it regression?

no

### Expected behavior

Simulators work smoothly and boot / open urls without crashing.

### Actual behavior

Attempts to boot simulators and navigate to a URL fail very frequently.

### Repro steps

Use the following script in a macOS 13 runner workflow:
`          open /Applications/Xcode_14.3.1.app/Contents/Developer/Applications/Simulator.app/
          phone1=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl create iPhone-hzso25lt6h9 com.apple.CoreSimulator.SimDeviceType.iPhone-14 com.apple.CoreSimulator.SimRuntime.iOS-16-4)
          echo "${phone1}"
          boot1=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl boot ${phone1})
          echo "${boot1}"
          phone2=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl create iPhone-sios1839ti com.apple.CoreSimulator.SimDeviceType.iPhone-14 com.apple.CoreSimulator.SimRuntime.iOS-17-0)
          echo "${phone2}"
          boot2=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl boot ${phone2})
          rm -r ~/Library/Caches/*
          rm -r ~/Library/Developer/CoreSimulator/Caches/*
          sleep 120
          nav1=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl openurl booted 'https://google.com')
          echo "${nav1}"
          echo "${boot2}"
          rm -r ~/Library/Caches/*
          rm -r ~/Library/Developer/CoreSimulator/Caches/*
          sleep 120
          nav2=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl openurl ${phone2} 'https://google.com')
          echo "${nav2}"`

**Comment by sergei-pyshnoi** [comment]
Hello @fpkamp . Thanks for your report. MacOS 13 in beta state for now so it can have some issues like this. We will investigate problem and return with feedback as we get any result.

**Comment by ilia-shipitsin** [comment]
@fpkamp , we are preparing new macos-13 image with xcode update. we are not certain on time framing, probably 2 weeks or so.

can you provide repro steps so we can test your scenario on updated image ?

**Comment by joshuapoq** [comment]
Crossposting my message here as it might relate: https://github.com/actions/runner-images/issues/8023#issuecomment-1661916389

> Our builds are now reaching an hour and timing out. It's taking from 2 to 11 minutes of time between each test bundle for our frameworks (24 test bundles). It was previously taking between 3 and 15 seconds.
I think this new deploy might be to blame as there were no issues on Monday. This is a considerable amount of added time.

**Comment by ilia-shipitsin** [comment]
@joshuapoq , can you help with repro steps (or link to repo/build) ?

**Comment by fpkamp** [comment]
@ilia-shipitsin sorry, I missed your ping from last week. The repro steps are in the original ticket - to clarify, the part I pasted in the repro steps is a bash script you should execute in your workflow. You may need to install the simulators before, 
so you want to first do 
xcodebuild -downloadPlatform iOS

**Comment by ilia-shipitsin** [comment]
hmm, is `open` a bash command ?

![image](https://github.com/actions/runner-images/assets/125650415/105f0e04-366f-41d7-9218-455488d6b77a)

**Comment by fpkamp** [comment]
Sorry, I meant just using a normal run action, feel free to just use this
` - name: Open devices 14.3
        run: |
          xcodebuild -downloadPlatform iOS
          open /Applications/Xcode_14.3.1.app/Contents/Developer/Applications/Simulator.app/
          phone1=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl create iPhone-hzso25lt6h9 com.apple.CoreSimulator.SimDeviceType.iPhone-14 com.apple.CoreSimulator.SimRuntime.iOS-16-4)
          echo "${phone1}"
          boot1=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl boot ${phone1})
          echo "${boot1}"
          phone2=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl create iPhone-sios1839ti com.apple.CoreSimulator.SimDeviceType.iPhone-14 com.apple.CoreSimulator.SimRuntime.iOS-17-0)
          echo "${phone2}"
          boot2=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl boot ${phone2})
          nav1=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl openurl booted 'https://google.com')

          echo "${nav1}"
          echo "${boot2}"
          nav2=$(/Applications/Xcode_14.3.1.app/Contents/Developer/usr/bin/simctl openurl ${phone2} 'https://google.com')
          echo "${nav2}"`
14.3.1 because it worked better, but obviously you can also use 15.0, I am currently running a build on the 13.5 image so I'll update soon as I get the results.

**Comment by ilia-shipitsin** [comment]
starting with yesterday we included simulators for XCode-15 as well, i.e. `xcodebuild-downloadAllPlatforms`, probably `xcodebuild -downloadPlatform iOS` can be omitted

**Comment by fpkamp** [comment]
Just to confirm - macos-13.5 is the expected input? I see that my job has been stuck in waiting for queue for the past 4 hours, but running macos13-xl was immediate (just to rule out something on my org end I haven't been notified about)

**Comment by ilia-shipitsin** [comment]
> Just to confirm - macos-13.5 is the expected input? I see that my job has been stuck in waiting for queue for the past 4 hours, but running macos13-xl was immediate (just to rule out something on my org end I haven't been notified about)

sorry, what do you mean by "input" ?

**Comment by fpkamp** [comment]
> > Just to confirm - macos-13.5 is the expected input? I see that my job has been stuck in waiting for queue for the past 4 hours, but running macos13-xl was immediate (just to rule out something on my org end I haven't been notified about)
> 
> sorry, what do you mean by "input" ?

is 
`runs-on: macos-13.5` the correct way to specify I want my workflow to [run on the image you deployed this week ](https://github.com/actions/runner-images/issues/7971#issuecomment-1660520377)?

**Comment by ilia-shipitsin** [comment]
> > > Just to confirm - macos-13.5 is the expected input? I see that my job has been stuck in waiting for queue for the past 4 hours, but running macos13-xl was immediate (just to rule out something on my org end I haven't been notified about)
> > 
> > 
> > sorry, what do you mean by "input" ?
> 
> is `runs-on: macos-13.5` the correct way to specify I want my workflow to [run on the image you deployed this week ](https://github.com/actions/runner-images/issues/7971#issuecomment-1660520377)?

no.... and yes ))

labels supported:

macos-11
macos-12
macos-13
macos-latest (alias for macos-12).

if you specify "macos-13.5" GHA will try to find your attached self hosted runner with that label.

**Comment by fpkamp** [comment]
> > > > Just to confirm - macos-13.5 is the expected input? I see that my job has been stuck in waiting for queue for the past 4 hours, but running macos13-xl was immediate (just to rule out something on my org end I haven't been notified about)
> > > 
> > > 
> > > sorry, what do you mean by "input" ?
> > 
> > 
> > is `runs-on: macos-13.5` the correct way to specify I want my workflow to [run on the image you deployed this week ](https://github.com/actions/runner-images/issues/7971#issuecomment-1660520377)?
> 
> no.... and yes ))
> 
> labels supported:
> 
> macos-11 macos-12 macos-13 macos-latest (alias for macos-12).
> 
> if you specify "macos-13.5" GHA will try to find your attached self hosted runner with that label.

Alright, I see, so can you please ELI5 how to use the aforementioned image?

**Comment by ilia-shipitsin** [comment]
aforementioned image is used automatically if you specify `runs-on: macos-13`, currently deployed image is used for that label. no way to choose another (or previously deployed) image.

**Comment by fpkamp** [comment]
@ilia-shipitsin okay so I finally ran it on the new version and looks pretty much the same in terms of stability.

**Comment by fpkamp** [comment]
hi there, was wondering if there's any updates? Is there a new version of the runner coming, maybe? Anything I can do to help?

**Comment by ilia-shipitsin** [comment]
Hey, @fpkamp , my access to test farm was restored today, I'll have a look

**Comment by ilia-shipitsin** [comment]
@stancho-stanchev , in current issue we investigate "Attempts to boot simulators and navigate to a URL fail very frequently."

if you observe another regression, please open an issue.

**Comment by ilia-shipitsin** [comment]
I acknowledge issue, @fpkamp , thank for repro steps.
I'm able to run simulators using your commands. iphone is starting, even it opens www.google.com, but everything is slow.
not yet sure on which side is an issue, is it simulator itself or something with our vm.

![image](https://github.com/actions/runner-images/assets/125650415/ead12679-5ebe-4447-b1f1-0b6cd6ced6ab)

**Comment by fpkamp** [comment]
@ilia-shipitsin cool, thank you. Maybe Arthur's observations here can help: [Ticket 2248137](https://support.github.com/ticket/2248137). And just to confirm - did you get any exceptions or errors while issuing the commands?
If that helps, I can also say that I have recently noticed that Safari driver started failing with a lot higher rate with some exotic error suggesting the server was already closed, but I wasn't able to find any more details. So I think the problem is in the VM.

**Comment by ilia-shipitsin** [comment]
@fpkamp , XCode 15 Beta 6 was just deployed, please test.

I did some dive on local installation. No conclusion yet, it does not work as I expect, but I'm not sure whether it is bug or feature.

**Comment by fpkamp** [comment]
@ilia-shipitsin 
I've noticed there are two XCode 15's in the /Applications folder:
Xcode_15.0.0.app
Xcode_15.0.app
Which one should I use? Which one is the most current beta?

**Comment by ilia-shipitsin** [comment]
those are identical.  I guess that Xcode_15.0.app is a symlink

**Comment by fpkamp** [comment]
sorry I have been meaning to update on my side, but I was fighting a different problem. I ran the same workflow and it looked the same except for failing on deleting the caches, hence I was going to run a workflow tomorrow where I wouldn't remove those, I'll do that and let you know, but not holding my breath.

**Comment by ilia-shipitsin** [comment]
it's utilizing all 8 cores when booting iPhone

![image](https://github.com/actions/runner-images/assets/125650415/90f9d3fe-63ab-4fc2-a5bd-b841614f2425)

**Comment by fpkamp** [comment]
I can confirm I see no improvement.

**Comment by ilia-shipitsin** [comment]
I narrowed it a little. 

```
2023-08-18 22:06:46.181524+0000 0x4a7a     Default     0x0                  1316   0    ReportCrash: (OSAnalytics) CFBundle URL refined <private> to <private>
2023-08-18 22:06:46.188679+0000 0x47e8     Default     0x0                  1316   0    ReportCrash: (OSAnalytics) CFBundle URL refined <private> to <private>
2023-08-18 22:06:46.189954+0000 0x4a7a     Default     0x0                  1316   0    ReportCrash: (OSAnalytics) CFBundle URL refined <private> to <private>
2023-08-18 22:06:46.274019+0000 0x4dd2     Default     0x0                  0      0    kernel: promotedcontentd[1895] Corpse failure, too many 6
2023-08-18 22:06:46.274028+0000 0x4dd2     Default     0x0                  0      0    kernel: Process[1895] crashed: promotedcontentd. Too many corpses being created.
```

"promotedcontentd" is a part of simulator. something makes it crashing.

good news - it does not crash on clean machine. I'll try to find what makes it crash

**Comment by fpkamp** [comment]
hi @ilia-shipitsin just wondering if you had more time to look into what was causing the crash?

**Comment by ilia-shipitsin** [comment]
it appeared to be more complex.

for some versions of XCode (like 15.0.0 Beta 5) it crashed, but for later (like Beta 8) it does not crash.
however, resource consumption is very high in all cases.

it does open url, but it takes quite a lot of time to complete.
we run on anka vm on top of mac mini, I'm trying to figure out what is bottleneck and whether we are supposed to be able to run 2-3 simulators at a time or not

**Comment by christiangal-indi** [comment]
Hi I have a different issue running my tests on the simulator, but it might be related to the slowness of the simulator as the same tests on Xcode14 run ok and it is random on the Xcode15.

UITests-Runner (9588) encountered an error (The test runner failed to initialize for UI testing. (Underlying Error: Timed out while loading Accessibility.))

**Comment by ilia-shipitsin** [comment]
@christiangal-indi , your issue may be either related or unrelated to this issue.

feel free to report your issue separately with repro steps. we can mention this issue if they are related

**Comment by joshuapoq** [comment]
@christiangal-indi - we're seeing exactly the same thing. After upgrading to Xcode 15 (and previously with this bug for Xcode 14) the iOS 17 simulators are completely terrible and unreliable for UI tests. The simulators take at least 10m before either starting or failing the build then our builds timeout (set to 1h) due to how slow they run (average build time was previously 20m).

**Comment by joshuapoq** [comment]
@christiangal-indi we're seeing the same thing. Upgraded to Xcode 15 (but it's the same problem described by this thread) that the iOS 17.0 runners are completely unusable. They're taking around 10m to launch and then either fail the build or run incredibly slowly causing our builds to timeout (which has a huge knock on effect).

This might be on Apple and related to: https://developer.apple.com/forums/thread/737695

We'll likely have to downgrade to Xcode 14 for validation lanes but use Xcode 15 for releases. Perhaps this will become more stable when the MacOS 14 images are released.

**Comment by NorseGaud** [comment]
Just wanted to include my 2-cents here: it's not even using any resources to make sense of why it's taking forever to boot.
<img width="1018" alt="Screenshot 2023-09-28 at 3 09 56 PM" src="https://github.com/actions/runner-images/assets/5896030/1a9f8041-f875-48c9-b5f7-a69c47f20ff0">

Lots of this log spam from the Console for the simulator: 
```
Connection for <APSConnection: 0x6000002802a0> differs from current queue!!
Unable to copy activation identity: _APSCopyActivationIdentity() failed. Device is probably unactivated. Do you need to hacktivate? {error: (null)}
<APSKeychainClientIdentityProvider: 0x60000332a280> SecItemDelete() failed: -25300
<APSUserCourier 0x1595061a0 development 0>: Terminating connection, deleting identity and clearing public token.
apsd (CFNetwork) Connection 4820: missing error, so heuristics synthesized error(1:53)
apsd (CFNetwork) Connection 4818: encountered error(1:53)
```

**Comment by wtto00** [comment]
It seems like it's an issue with Xcode 15. I ran a sample example: https://github.com/wtto00/test-github-action/actions/runs/6397195336

```shell
sudo xcode-select -s "/Applications/Xcode_15.0.app"
IPHONE_MODEL="CA968C7D-0661-4517-9E8F-06A54215C5FD" # iPhone 15 with iOS 17.0 on xcode_15_macos_13

# IPHONE_MODEL="E8C11E27-784B-4FC0-AB92-D79C0307B337" # iPhone 14 with iOS 16.4 on xcode_14_macos_13
# IPHONE_MODEL="92DF7F17-8DA7-400E-975F-7AA7C5004EA0" # iPhone 14 with iOS 16.2 on xcode_14_macos_12

rm -rf ~/Library/Caches/*
rm -rf ~/Library/Developer/CoreSimulator/Caches/*

open -a Simulator

xcrun simctl list

xcrun simctl boot "$IPHONE_MODEL"

echo 'wait for booted'
while true; do
  status=$(xcrun simctl list | grep "Phone:" | grep "$IPHONE_MODEL" | awk '{print $5}')
  echo $(xcrun simctl list | grep "Phone:" | grep "$IPHONE_MODEL")
  echo "status: $status"
  if [[ "$status" == "(Booted)" ]]; then
    break
  fi
  sleep 1
done

echo 'openurl'
xcrun simctl openurl $IPHONE_MODEL "https://google.com"
```
When running `iPhone 15 with iOS 17.0 on xcode_15_macos_13`, it will throw an error like this:
```
An error was encountered processing the command (domain=NSPOSIXErrorDomain, code=60):
Simulator device returned an error for the requested operation.
Operation timed out
Underlying error (domain=NSPOSIXErrorDomain, code=60):
	The operation couldn’t be completed. Operation timed out
	Operation timed out
```
But the other two cases will not throw an error.

**Comment by NorseGaud** [comment]
Apple just released `iOS 17.0.1 Simulator Runtime` so maybe they've included some fixes for these problems.

**Comment by NorseGaud** [comment]
Xcode 15.1 release notes

[Resolved Issues](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_1-release-notes#Resolved-Issues)

Fixed: Creating a new visionOS App target sets the Application Scene Manifest (Generation) target build setting to Yes by default. Changing to Application Scene Manifest using the Info dictionary editor doesn’t remove the build setting “Application Scene Manifest (Generation)” as expected, which will override changes with the default value. (109428090)

**Fixed: Executing Unit/UI tests from Xcode on the iOS Simulator takes an extended time to launch on first run of the suite. (115187363) (110330776)**

Fixed: Swift macro definitions from the macOS/iOS/tvOS/watchOS SDKs are not available in Swift Playgrounds. (112122752)

Fixed: When building with the visionOS SDK, TargetConditionals.h incorrectly sets TARGET_OS_IOS to 1 instead of 0. (112716373)

**Fixed: The simulator may crash when opening Settings or Action Button settings on iPhone 15 Pro devices. (115388496)**

**Comment by NorseGaud** [comment]
Just ran a test on Xcode 15.1 beta 15C5028h AND 17.0.1, but Simulator boot times are still super slow or don't load at all :(

**Comment by ilia-shipitsin** [comment]
hello, @NorseGaud , @wtto00 , @joshuapoq , @fpkamp , @stancho-stanchev 

I've made some investigation, so far it looks like 

1. simulator is started
2. it creates additional multicast targets
3. mDNSResponder on the host is happy to log those messages
4. IO is wasted by logging/mDNS

mDNSResponder on the host seem to be nonneccessary and it is a leftover os using desktop mac mini.
in my tests I gained some speedup by

```
   sudo launchctl unload -w /System/Library/LaunchDaemons/com.apple.mDNSresponder.plist
   sudo launchctl unload -w /System/Library/LaunchDaemons/com.apple.mDNSresponderHelper.plist
   sudo launchctl unload -w /System/Library/LaunchDaemons/com.apple.airportd.plist
```

can you please try those commands in your pipelines to check

1. whether it will bring some side effect
2. whether it provides speedup or not

thanks!

**Comment by mikehardy** [comment]
@ilia-shipitsin - I tried that as I'm very interested in restoring performance to the macOS-12 levels, but it made DNS in general non-functional. Test results indicate that after running that it is no longer possible to resolve anything so git checkout fails.

Perhaps it is possible to carefully order *when* these commands are run such that the only activity after running them is machine-local but that seems like a fragile idea so I won't pursue it.

Runs linked to failure line, "macOS-13-runner speedup" steps are just above them:
https://github.com/invertase/react-native-firebase/actions/runs/6969529848/job/18965658862#step:3:58
https://github.com/invertase/react-native-firebase/actions/runs/6969529845/job/18965654159#step:6:59

**Comment by ilia-shipitsin** [comment]
@mikehardy , thank for your feedback.
can you try unloading diagnosticd instead ?

```
sudo launchctl unload -w /System/Library/LaunchDaemons/com.apple.diagnosticd.plist
```

(also, I've found that simulator has its own diagnosticd, I'm looking at it)

**Comment by ilia-shipitsin** [comment]
@mikehardy , I'll try different things on https://github.com/invertase/react-native-firebase

**Comment by mikehardy** [comment]
@ilia-shipitsin hmm not easily. I think you could change one line that specifies the simulator to something available on both and it would work though?

https://github.com/invertase/react-native-firebase/blob/main/tests/package.json#L66

https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md#installed-simulators
https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md#installed-simulators

"iPhone 14" is *allegedly* available for the value of Xcode that is "xcode-current" on macOS12 and macOS13, although that wasn't working for me for some reason on Xcode 15 - I don't recall trying too hard to investigate or fix it though

If that doesn't work, could easily add a matrix check in the workflow yaml you are working with such that you execute a "tests:ios:test-cover" for macOS-13 and a new "tests:ios:test-cover-macos12" script you [add here](https://github.com/invertase/react-native-firebase/blob/main/package.json#L47) that references a new "ios.sim.debug.macos12" stanza you [add here](https://github.com/invertase/react-native-firebase/blob/main/tests/package.json#L60) with iPhone 14 for macOS12

a little bit of a detour vs your goal here but I think that would work well and be quick to do

**Comment by ilia-shipitsin** [comment]
@mikehardy , thank for the tip. I'm able to run "iPhone 14" on both macos-12 and macos-13. give me some time to compare results.

actually, macos-12 runs test in 7 min while macos-13 need ~50 min even with "yeetd" enabled. no idea why, but we can try to find out

**Comment by mikehardy** [comment]
> @mikehardy , thank for the tip. I'm able to run "iPhone 14" on both macos-12 and macos-13. give me some time to compare results.
> 
> actually, macos-12 runs test in 7 min while macos-13 need ~50 min even with "yeetd" enabled. no idea why, but we can try to find out

macos-13 has been about that painful, yes 😂. Appreciate that you're examining, I believe everyone will appreciate any progress!

**Comment by ilia-shipitsin** [comment]
@mikehardy , what caught my eyes is really huge simulator log. I noticed that tests run in debug mode. can we try to either disable logging or reduce them to release ?

I collected logs from macos-12 and macos-13 and parsed them by Microsoft LogParser using the following query

```
SELECT COUNT(*) AS Total, SUBSTR(EXTRACT_PREFIX(Field1,0,'['),24) AS Service FROM 'C:\i\simulator-log\13\simulator.log'
WHERE Field1 like '2023%'
GROUP BY Service
ORDER BY Total DESC
```

![image](https://github.com/actions/runner-images/assets/125650415/51efaf64-53ab-481e-b241-84f05c992b71)

![image](https://github.com/actions/runner-images/assets/125650415/d82eaadf-1fb0-47ea-ae2c-d6f29141c8c9)



something interesting with SpringBoard. what's that )) ? 

(well, I suspect it might be a regression on XCode itself or some side effect of running simulators on virtualized hardware, @NorseGaud  do you have any idea why SpringBoard can be that noisy under anka platform maybe ?)

**Comment by ilia-shipitsin** [comment]
the most noisy `SpringBoard` activity is `[com.apple.xpc:connection]`

![image](https://github.com/actions/runner-images/assets/125650415/60fdd2ac-0fbc-47d2-b067-b85e78cc5191)

**Comment by ilia-shipitsin** [comment]
as for "xpc connection" .... 


![image](https://github.com/actions/runner-images/assets/125650415/42318598-7e9f-4917-a136-fea6307700dd)

**Comment by mikehardy** [comment]
> @mikehardy , what caught my eyes is really huge simulator log. I noticed that tests run in debug mode. can we try to either disable logging or reduce them to release ?

Well, anything is possible @ilia-shipitsin :-) - my use case for the simulator.log is so that I can troubleshoot things like this when they go wrong though - it is not always our app, sometimes it is some rogue networking thing though, and the simulator.log is how it is discovered - as you've done (awesome by the way)

I'm willing to try anything but I don't know how I'll do a more efficient job than you are already doing on the fork. From the perspective of our workflows the simulator.log is not used as any sort of end product it is just for troubleshooting, so you could disable the simuator.log capture / zip / upload entirely and our workflow would still be doing it's job (assuming there wasn't an app crash where I needed to grab the stack trace...)

I'm going to guess though, disabling the simulator log will provide a substantial speedup by removing IO, but if you assume that the job is compute-bound and the IO is streaming/buffered then my guess is the macos13/Xcode15/ios17 slowdown we are all seeing is because the Simulator's networking/springboard subsysytem is doing something terrible that is *also* CPU heavy, so the speedup will not be nearly enough to get back to macOS12/Xcode14/ios16 levels

https://developer.apple.com/documentation/xpc/1448777-xpc_connection_cancel - XPC is some interprocess communication, not sure which processes are trying to communicate and/or why it's failing but something is definitely not happy on these new versions of the  iOS simulator stack

**Comment by ilia-shipitsin** [comment]
@mikehardy , we did more investigation (will provide result later). also, we tried to run react-native-firebase on arm64 (also virtualized) runner

https://github.com/bbq-beets/react-native-firebase/actions/runs/7030335355/job/19129684183

as far as I understand, arm64 does not support nested virtualization, but it should only affect android simulators, not iOS. 

can we add some debug to find why `09:45:59.561 detox[12510] ERROR: [APP_UNREACHABLE] Detox can't seem to connect to the test app(s)!` ?

**Comment by mikehardy** [comment]
@ilia-shipitsin sorry for the delay!

The test app connection is the last step in a wobbly tower of things that have to go correctly, it happens when the test infrastructure correctly requests Simulator start, the operating system has started the simulator correctly, the simulator has booted completely and the test infrastructure can see it is up, then the test infrastructure has correctly loaded the app on to the simulator and asked it to start and finally the app itself has started completely and fetched the javascript bundle from the bundle server, loaded it and started executing

Determining why it did not start is what the simulator.log running in Debug is for :-)

I look for a few markers to see what stage things got to. One is the case-insensitive string "crashlytics" - if this never shows up, the app native code never booted and ran so as a binary search, we're looking at simulator startup failure or app load/start failure (where failure may be timeout / just took too long)

If "crashlytics" shows up then things should start happening pretty quickly and I search for the app name and/or "react" and/or "rnfb" to see if the javascript bundle loaded and app components started loading and getting chatty

My guess is this was just more poor performance (a 2GB+ log file! ouch) meaning things were happening slowly enough that the test infrastructure considered it a failure and timed it out

**Comment by jeanregisser** [comment]
Hey, I'm dealing with similar issues using: `macos-13-xlarge` (running on M1).

I was unable to get Detox to work using multiple workers.
Apparently it fails too boot the simulators, or they are really slow.
I see a lot of `14:34:36.027 detox[60482] i Error: Unable to update lock within the stale threshold` right after starting the detox test.
See the [full logs](https://github.com/valora-inc/wallet/actions/runs/7061213984/job/19222385590).

Then I tried using [a single worker](https://github.com/valora-inc/wallet/pull/4570/commits/f44f52fdf42d41b6f45d42d231e22dfd55837a63).
And things got better. But it was still quite slow and it reached the 45 mins timeout I had set for that part of the workflow.
See the [full logs](https://github.com/valora-inc/wallet/actions/runs/7061213984/job/19222385590).

This was using Xcode 14.3.1 and iOS 16.4 simulators.

Is there a limitation on how many simulators can be run?

Anyway, thanks for all the useful info in this thread. I'm gonna try a few more things. Let me know if there's anything I can do to help. 🙏

**Comment by NorseGaud** [comment]
Quick update from my testing with `react-native-firebase`:

If I run this manually, I can choose the GPU type in the Simulator menu and see different times with a slight performance improvement using Integrated GPU.

```
Integrated GPU in Simulator:
SIMCTL_CHILD_GULGeneratedClassDisposeDisabled=1 ./node_modules/.bin/nyc yarn   13.97s user 2.49s system 6% cpu 4:29.95 total

Discrete GPU in Simulator:
SIMCTL_CHILD_GULGeneratedClassDisposeDisabled=1 ./node_modules/.bin/nyc yarn   14.56s user 3.61s system 5% cpu 5:07.03 total
```

Not much of a difference, but it's something to note as a possibility for improvement.

The `messaging()` tests timeout in any kind of virtualization, but seems like all other tests work fine. Maybe relevant for @mikehardy. I collected the Console logs from macOS while the test was seemingly hanging.

[console-messaging-full.log](https://github.com/actions/runner-images/files/13687313/console-messaging-full.log)
[console-messaging-errors-only.log](https://github.com/actions/runner-images/files/13687314/console-messaging-errors-only.log)

I don't know what `firebase.messaging` in the test is actually doing, but I do think there may be something from the Console logs that the developers could see that helps us pinpoint for Apple what's wrong.

**Comment by mikehardy** [comment]
Interesting @NorseGaud - do you  have a workflow run URL you can point me to where you extracted those logs? Or could you specify the execution environment? I'm to go on a hunch that this was an Apple Silicon machine of some sort?

That is a difference (intel silicon mac vs apple silicon mac) that varies the messaging testing as apple silicon macs with latest emulators can actually generate APNS tokens and receive APNS messages and we attempt to test that if it's recent enough simulator on apple silicon. It should work of course, but - for the purposes of the testing here, we are an open source repository limited to running on the currently available intel silicon runners, so I don't think it's germane to the current focus here unfortunately

**Comment by NorseGaud** [comment]
> do you have a workflow run URL you can point me to where you extracted those logs

I don't :( I setup the project manually in a VM of several virtualization tools and got the same results across them all.

Specs for the logs provided:

- Xcode 15.1 / iPhone 14 / 16.4
- macOS 14.2
- Silicon arch

Understood about not testing on Arm, though the result is exactly the same on macOS 13-14 and Xcode 15.x, regardless of the architecture. It's just faster and easier for me to test on ARM right now :)

Unless there is something in these logs that indicates a problem with a service that VMs don't have or isn't working right, I do worry we won't be able to do much until we can describe in detail what Apple has to fix :(

**Comment by NorseGaud** [comment]
I am unable to reproduce issues anymore with iOS 17.2. The simulators load just fine (but are still a bit heavy on usage) for my test apps and no longer hang.

**Comment by mikehardy** [comment]
Good to know @NorseGaud but unfortunately does not look like that's available in hosted runners yet https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md#xcode

**Comment by NorseGaud** [comment]
Sorry @mikehardy , I was speaking about simulator booting issues. The issues with your tests persist AFAIK.
